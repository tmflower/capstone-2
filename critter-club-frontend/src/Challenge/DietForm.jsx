import { useState } from 'react';
import { Paper, FormLabel, RadioGroup, Radio, Typography, Button } from '@mui/material';
import { theme } from '../theme/theme';

/** DietForm renders a question with radio button answers */

export function DietForm({commonName, diet, message, points, setPoints, numQuestions, setNumQuestions }) {

    // Display form with no default answers and no initial message
    const initialState = {animalDiet: ''};
    const [formData, setFormData] = useState(initialState);
    const [feedback, setFeedback] = useState('');

    // Deconstruct formData properties
    const { animalDiet } = formData;

    // Set formData to user's input
    const handleChange = (e) => {
        const { name, value } = e.target;
		setFormData((formData) => 
            ({ ...formData, [name]: value }));
    }

    // compare user's selection to animal data to check if correct; provide corresponding feedback message
    // update user's points
    // increment numQuestions; when numQuestions is >=3, user can submit answers and earn badge
    const handleSubmit = (e) => {
        e.preventDefault();
        if (animalDiet === diet) {
            setFeedback(message.correct);
            setPoints(points+=10);
            setNumQuestions(numQuestions+1);
        }
        else setFeedback(message.incorrect);
    }

    // clear form responses & message when user clicks "Clear answer" button
    const handleReset = (e) => {
        setFormData(initialState);
        setFeedback('');
    }

    return (
        <Paper
        elevation={8}
        sx={{ padding: 3, mt: 3 }}>
        <form onSubmit={handleSubmit}>
            <fieldset>
            <RadioGroup>
                <FormLabel sx={{ fontSize: '1.5rem', fontFamily: 'Lexend Deca, Arial',}}>
                    <legend>Which term best describes the {commonName.toLowerCase()}'s diet?</legend>
                </FormLabel>
            
            <label htmlFor='Carnivore'>
            <Radio 
                name="animalDiet"
                value="Carnivore"
                id="Carnivore"
                onChange={handleChange}
                checked={animalDiet === "Carnivore"}
            />
            Carnivore
            </label>
            <label htmlFor='Herbivore'>
            <Radio 
                name="animalDiet"
                value="Herbivore"
                id="Herbivore"
                onChange={handleChange}
                checked={animalDiet === "Herbivore"}
            />
            Herbivore
            </label>
            <label htmlFor="Omnivore">
            <Radio
                name="animalDiet"
                value="Omnivore"
                id="Omnivore"
                onChange={handleChange}
                checked={animalDiet === "Omnivore"}
            />
            Omnivore
            </label>
            <div>
                { !feedback ? 
                <Button type="submit">Check answer</Button>
                : null }
                { feedback === message.incorrect ?
                <Button id="alt-button" type="reset" onClick={handleReset}>Clear answer</Button>
                : null }
            </div>
            <Typography id="quiz-feedback" variant="h5" sx={{ fontFamily: theme.typography.primary, textAlign: 'center' }}>{feedback}</Typography>
            </RadioGroup>
            </fieldset>
        </form>
        </Paper>
    )

}