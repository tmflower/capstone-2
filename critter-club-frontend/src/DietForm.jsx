import { useState } from 'react';

export function DietForm({commonName, diet}) {

    const [formData, setFormData] = useState('');
    const { animalDiet } = formData;

    const handleChange = (e) => {
        const { name, value } = e.target;
		setFormData((formData) => 
            ({ ...formData, [name]: value }));
    }

    const handleSubmit = (e) => {
        // compare user's selection to animal data to check if correct
        // TODO: add functionality for user feedback
        e.preventDefault();
        if (animalDiet === diet) {
            console.log("Correct on diet!")
        }
        else console.log("sorry, try again");
    }

    const handleReset = (e) => {
        setFormData('');
    }

    return (
        <form onSubmit={handleSubmit}>
            <fieldset>
            <legend>Which term best describe's the {commonName.toLowerCase()}'s diet?</legend>
            <label>
            <input 
                type="radio"
                name="animalDiet"
                value="Carnivore"
                onChange={handleChange}
                >
            </input>
            Carnivore
            </label>
            <label>
            <input 
                type="radio"
                name="animalDiet"
                value="Herbivore"
                onChange={handleChange}
                >
            </input>
            Herbivore
            </label>
            <label>
            <input 
                type="radio"
                name="animalDiet"
                value="Omnivore"
                onChange={handleChange}
                >
            </input>
            Omnivore
            </label>
            <div>
                <button type="submit">Check answer</button>
                <button type="reset" onClick={handleReset}>Clear answer</button>
            </div>
            </fieldset>
        </form>
    )

}