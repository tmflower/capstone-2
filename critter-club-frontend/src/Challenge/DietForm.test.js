import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { DietForm } from './DietForm';
import UserContext from '../userContext';
import userEvent from "@testing-library/user-event";

const currentUser = {user: { username: "Tiny", id: 1, password: 'abc123$%^', points: 40, userBadges: ['Koala', 'Jellyfish', 'Painted Turtle'], parentId: 1 }}

const message = {
  correct: "‚≠êÔ∏è‚≠êÔ∏è‚≠êÔ∏è You got it! ‚≠êÔ∏è‚≠êÔ∏è‚≠êÔ∏è",
  incorrect: "üòßüòßüòß Sorry, try again. üòßüòßüòß"
};

const setPoints = () => {
  return;
}

const setNumQuestions = () => {
  return;
}

// smoke test
test("renders without crashing", () => {
    render(
    <MemoryRouter>
      <UserContext.Provider value={currentUser}>
        <DietForm 
          commonName="Hippopotamus"
          message="some message"/>
      </UserContext.Provider>      
    </MemoryRouter>);
  });

// snapshot test
test("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={currentUser}>
          <DietForm 
            commonName="Hippopotamus"
            message="some message"/>
        </UserContext.Provider>        
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

// tests for user interactions
test("user submits wrong answer", async () => {
  const user = userEvent.setup();
  const setPoints = () => {return}
  render(  
    <MemoryRouter>
      <UserContext.Provider value={currentUser}>
        <DietForm 
        commonName="Hippopotamus"
        message={message}
        diet="Herbivore"
        setPoints={setPoints}
        />
      </UserContext.Provider>      
    </MemoryRouter>);

  const wrong = screen.getByLabelText("Carnivore");
  expect(wrong).not.toBeChecked();
  await user.click(wrong);
  expect(wrong).toBeChecked();
  await user.click(screen.getByText("Check answer"));
  expect(screen.getByText("üòßüòßüòß Sorry, try again. üòßüòßüòß")).toBeInTheDocument(); 
});

test("user submits correct answer", async () => {
  const user = userEvent.setup();

  render(  
    <MemoryRouter>
      <UserContext.Provider value={currentUser}>
        <DietForm 
        commonName="Hippopotamus"
        message={message}
        diet="Herbivore"
        setPoints={setPoints}
        setNumQuestions={setNumQuestions}
        />
      </UserContext.Provider>      
    </MemoryRouter>);

  const right = screen.getByLabelText("Herbivore");
  expect(right).not.toBeChecked();
  await user.click(right);
  expect(right).toBeChecked();
  await user.click(screen.getByText("Check answer"));
  expect(screen.getByText("‚≠êÔ∏è‚≠êÔ∏è‚≠êÔ∏è You got it! ‚≠êÔ∏è‚≠êÔ∏è‚≠êÔ∏è")).toBeInTheDocument(); 
});