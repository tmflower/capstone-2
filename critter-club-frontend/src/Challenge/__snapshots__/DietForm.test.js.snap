// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 css-2e4xct-MuiPaper-root"
  >
    <form>
      <fieldset>
        <div
          class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            class="MuiFormLabel-root MuiFormLabel-colorPrimary css-tshmje-MuiFormLabel-root"
          >
            <legend>
              Which term best describes the hippopotamus's diet?
            </legend>
          </label>
          <label
            for="Carnivore"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Carnivore"
                name="animalDiet"
                type="radio"
                value="Carnivore"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Carnivore
          </label>
          <label
            for="Herbivore"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Herbivore"
                name="animalDiet"
                type="radio"
                value="Herbivore"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Herbivore
          </label>
          <label
            for="Omnivore"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                id="Omnivore"
                name="animalDiet"
                type="radio"
                value="Omnivore"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Omnivore
          </label>
          <div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Check answer
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-18czui-MuiTypography-root"
            id="quiz-feedback"
          />
        </div>
      </fieldset>
    </form>
  </div>
</DocumentFragment>
`;
